### ISSUE TITLE
Train station inventories are only populated when there is a docked carriage

### ISSUE THREAD

**Issue by cosmopetrich** [porisius/FicsitRemoteMonitoring#101]
As originally noted by @featheredtoast in [this Discord thread](https://discord.com/channels/1142919853053841488/1186378513427730502), train station inventories will be shown as empty (i.e. with no platforms) unless there is a docked train at the platform.

This appears to be the result of FRM using `mDockedPlatformList`: https://github.com/porisius/FicsitRemoteMonitoring/blob/47269f391d45337da3ee4b27c69e469c1e31ba71/Source/FicsitRemoteMonitoring/Private/FRM_Trains.cpp#L144

The behaviour of this function appears to be expected. According to [a note in the SML source](https://github.com/satisfactorymodding/SatisfactoryModLoader/blob/04025f8209079ee869269fa3a64932fa3cde2b4b/Source/FactoryGame/Public/Buildables/FGBuildableRailroadStation.h#L137):

> When docked, this station will fill this array with every potential platform in its tail. 1 for each train segment

It would be good to return the inventory regardless of whether a train is docked, and regardless of whether the train is long enough to have a carriage at each platform of the station. To do so, I think it may be necessary to build a list of platforms by starting with [GetStationOutputConnection](https://github.com/satisfactorymodding/SatisfactoryModLoader/blob/04025f8209079ee869269fa3a64932fa3cde2b4b/Source/FactoryGame/Public/Buildables/FGBuildableRailroadStation.h#L56) and working along the connections=, resolving each with `GetPlatformOwner`. There may be an easier way, however.

Note that the original Discord report also mentioned duplicated platforms. I haven't seen that in my own testing. It's possibly my setup didn't trigger it, or that a newer version of FRM has fixed that issue.

**Comment by cosmopetrich** [comment]
I've been poking at this in spare moments. While the underlying problem isn't difficult, this is the first time I've had to deal with AccessTransformers and UE's reflection system so it hasn't been fast going. Lots of cases of "operating system error" or spurious segfaults. I'll keep poking at it regardless.

**Comment by featheredtoast** [comment]
my comment about duplicate platforms wasn't an FRM bug exactly, just an observation in how a train station may look at a platform.

As it is, interested parties may overcount things like items in inventory, or power use if connected platforms are counted for every station.

I'm using it to calculate a train system's power: with every station and platform connected to the grid, what is the current power draw and max power draw possible if every platform in the system were to be loading/unloading at once? Right now i have to tally:

```
for each station within a given power grid:
  for each platform connected to that station:
    grab current power use and max power use. Add up to a total.
```

This is easy to overcount things if you have multiple stations connected to the same platform - that platform's power use will be doublecounted in this calculation. A better method (not currently query-able in FRM) is:

```
for each station AND platform within a power grid:
  grab current power use and max power use. Add up to a total.
```
Platforms in this case are only counting a single time.

It might be helpful to publish these power metrics separately, perhaps at some point in getPower.

For reference, I'm gunning towards a power breakdown so one can easily discover how power is being used within a circuit and a breakdown across the factory (production buildings, trains, train stations, truck stations, sinks, teleporters, hypertubes, pumps, etc)

**Comment by cosmopetrich** [comment]
That makes sense. While I'm currently immediately trying to fix the "missing data when not docked" issue, once that's done perhaps it would be good to see if we can tweak the output schema for `getTrainStation` to make it easier for clients to use.

Regarding the issues I've been having solving the not-docked issue, [I described the problem I'm having in the modding Discord](https://discord.com/channels/555424930502541343/862002356626128907/1297401995002642493) but didn't get much in the way of replies to that or my bump the next day. So I'm not sure if I'm going about it the wrong way, I've hit a bug, or it's just my general lack of experience. It's quite annoying.

A site-wide code search on Github didn't really turn up any cases where people are using the same methods I'm having problems with, and even guessing at some mods which might be trying to do something similar (Ficsit Networks came to mind) and poking through their code didn't give me anything to go on.

**Comment by cosmopetrich** [comment]
Regarding the problem I mentioned above, there's a more thorough example of what I'm seeing in https://github.com/cosmopetrich/FicsitRemoteMonitoring/blob/a5e7b6860e443d8ae67d303f13b72d76bd6c1ff8/Source/FicsitRemoteMonitoring/Private/FRM_Trains.cpp.

The condition at https://github.com/cosmopetrich/FicsitRemoteMonitoring/blob/a5e7b6860e443d8ae67d303f13b72d76bd6c1ff8/Source/FicsitRemoteMonitoring/Private/FRM_Trains.cpp#L159-L167 will always trigger, even if there is definitely an subsequent platform connection, as shown by the fact that the not-is-connected test above does not evaluate true.

For example, creating a station with **no** platforms will cause `LogFRMDebug: Platform connection was not connected to another connection` to be written to `%localappdata%/FactoryGame/Saved/Logs/FactoryGame.log`, while creating a station with one or more platforms will instead trigger `LogFRMDebug: Next platform connection was NULL`.

I also tried a couple of other ways of getting the connection, as described in the gist I linked in the SML discord - https://gist.github.com/cosmopetrich/67c8a5b85d979861062538df8e4ec5fc - with the same result.

If for whatever reason it does end up not being feasible to find the attached platforms then [there was a discussion in the Discord](https://discord.com/channels/1142919853053841488/1142921725269524643/1297679649261883434) where @featheredtoast noted that a separate array of all platforms regardless of station would be just as useful. For my part I do think having the platforms associated with the station is preferable and I think it's more likely that there's a way to do it that I'm missing, but at least we've got another option.

tl;dr I don't know what I'm doing wrong so now I'm going to go sit in the shower and weep.

**Comment by cosmopetrich** [comment]
There's another commit regarding this in dev, 3a985d3.

**Comment by cosmopetrich** [comment]
It looks like [SML commit#3bbfe48](https://github.com/satisfactorymodding/SatisfactoryModLoader/commit/3bbfe48d2d041d76b2001e5f12a40a9d96040829) fixed the behaviour of `GetPlatformOwner()`, and so the code I linked above in https://github.com/porisius/FicsitRemoteMonitoring/issues/101#issuecomment-2425647118 now works.

I'll admit to being just a little tilted that necessary changes for the new version of SF are stuck in the SML dev branch despite the docs indicating that mod development should be done against main.
